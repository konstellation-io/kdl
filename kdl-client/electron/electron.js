const electron = require('electron');
const app = electron.app;
const BrowserWindow = electron.BrowserWindow;

const path = require('path');
const isDev = require('electron-is-dev');
const { autoUpdater } = require('electron-updater');

const SPLASH_SCREEN_MIN_TIME = 2000;
const SPLASH_SCREEN_SIZE = {
  width: 600,
  height: 400,
};
const WINDOW_SIZE = {
  width: 900,
  height: 680,
};

let mainWindow;
let splashScreen;

function showSplashScreen() {
  splashScreen = new BrowserWindow({
    ...SPLASH_SCREEN_SIZE,
    frame: false,
    backgroundColor: '#060b15',
  });

  splashScreen.loadURL(
    `file://${path.join(__dirname, '../public/splash_screen.html')}`
  );
}

function closeSplashScreen() {
  splashScreen.close();
  splashScreen = null;
}

function createMainWindow() {
  let mainWindowReady = false;

  mainWindow = new BrowserWindow({
    ...WINDOW_SIZE,
    show: false,
    webPreferences: {
      nodeIntegration: true,
    },
    transparent: true,
    // frame: false,   // TODO: set this to true after creating a custom title bar
    titleBarStyle: 'hidden',
  });

  mainWindow.loadURL(
    isDev
      ? 'http://localhost:3000'
      : `file://${path.join(__dirname, '../build/index.html')}`
  );

  if (isDev) {
    mainWindow.webContents.openDevTools();
  }

  mainWindow.once('ready-to-show', () => {
    if (!mainWindow) {
      throw new Error('"mainWindow" is not defined');
    }

    mainWindowReady = true;
  });

  // This will wait at least for two seconds before showing the main window.
  const showMainWindowInterval = setInterval(() => {
    if (mainWindow && mainWindowReady) {
      clearInterval(showMainWindowInterval);

      closeSplashScreen();

      mainWindow.show();
      mainWindow.focus();
    }
  }, SPLASH_SCREEN_MIN_TIME);

  mainWindow.on('closed', () => (mainWindow = null));
}

function createWindow() {
  showSplashScreen();
  createMainWindow();
}

// when the app is loaded create a BrowserWindow and check for updates
app.on('ready', function () {
  createWindow();
  if (!isDev) autoUpdater.checkForUpdates();
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  mainWindow && createWindow();
});

// when the update has been downloaded and is ready to be installed, notify the BrowserWindow
autoUpdater.on('update-downloaded', () => {
  mainWindow.webContents.send('updateReady');
});

// when receiving a quitAndInstall signal, quit and install the new version ;)
electron.ipcMain.on('quitAndInstall', () => {
  autoUpdater.quitAndInstall();
});
