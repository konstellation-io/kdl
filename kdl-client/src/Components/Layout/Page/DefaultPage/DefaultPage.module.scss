@import 'Styles/colors';
@import 'Styles/variables';
@import 'Styles/mixins';

$sidebar-width: 40 * $grid-unit;

.container {
  position: relative;
  display: flex;
  width: 100%;
  height: 100%;
  background-color: palette(base, 700);
}

.actions div {
  padding: 0;
  flex-direction: row;
}

.sidebar {
  width: $sidebar-width;
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  background-color: palette(base, 500);
  padding: 10 * $grid-unit 5 * $grid-unit 0 3 * $grid-unit;

  h1 {
    text-align: left;
  }

  .subtitle {
    margin-top: 3 * $grid-unit;
  }

  .bottom {
    margin-top: auto;
  }
}

.content {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  max-width: 1440px - $sidebar-width;
  background-color: palette(base, 800);
  overflow: auto;
}

.children {
  flex-grow: 1;
}
