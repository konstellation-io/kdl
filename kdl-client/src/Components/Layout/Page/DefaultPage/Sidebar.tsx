import React, { useRef } from 'react';

import styles from './DefaultPage.module.scss';

export let bottomRef: React.MutableRefObject<HTMLDivElement | null>;
export let topRef: React.MutableRefObject<HTMLDivElement | null>;

type Props = {
  title: string;
  subtitle?: string;
};
function Sidebar({ title, subtitle }: Props) {
  bottomRef = useRef<HTMLDivElement>(null);
  topRef = useRef<HTMLDivElement>(null);

  return (
    <div className={styles.sidebar}>
      <h1>{title}</h1>
      {subtitle && <h3 className={styles.subtitle}>{subtitle}</h3>}
      <div className={styles.top} ref={topRef} />
      <div className={styles.bottom} ref={bottomRef} />
    </div>
  );
}

export default Sidebar;
