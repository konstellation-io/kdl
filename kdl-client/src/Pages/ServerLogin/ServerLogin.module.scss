@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  width: 100%;

  .information {
    padding: 8 * $grid-unit 8 * $grid-unit 6 * $grid-unit 8 * $grid-unit;
    background-color: palette(base, 700);
  }
  .form {
    padding: 4 * $grid-unit 8 * $grid-unit 8 * $grid-unit 8 * $grid-unit;
  }

  .title {
    @include font-caption;
    color: font-color(regular);

    margin: 0 0 2 * $grid-unit 0;
  }

  .name .value {
    @include font-subheader;
    color: font-color(light);
    text-transform: uppercase;

    margin: 0;
  }

  .url .value {
    @include font-body;
    color: font-color(lowlight);

    text-overflow: ellipsis;
    overflow: hidden;
    margin: 0 0 4 * $grid-unit 0;
  }

  .formEmail {
    margin: 0 auto;
  }

  .confirmation {
    @include font-caption;
    color: palette(highlight);
    text-transform: uppercase;

    display: flex;
    align-items: center;
    margin: 0;
    height: 0;
    overflow: hidden;

    transition: margin ease 0.3s, height ease 0.4s;

    &.show {
      margin: 2 * $grid-unit 0 3 * $grid-unit 0;
      height: 6 * $grid-unit;
    }

    .icon {
      width: 4 * $grid-unit;
      flex-shrink: 0;
    }
  }
}

.cancelButton {
  width: 12 * $grid-unit;
}

.loginButton {
  width: 25 * $grid-unit;
}
