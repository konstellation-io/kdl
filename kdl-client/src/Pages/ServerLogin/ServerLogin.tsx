import { Button, CHECK, TextInput } from 'kwc';
import ROUTE, { RouteServerParams } from 'Constants/routes';
import React, { useEffect, useRef, useState } from 'react';

import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import EmailIcon from '@material-ui/icons/Mail';
import cx from 'classnames';
import { ipcRenderer } from 'electron';
import styles from './ServerLogin.module.scss';
import { useForm } from 'react-hook-form';
import { useParams } from 'react-router-dom';
import useServers from 'Hooks/useServers';

type ServerLoginResponse = {
  success: boolean;
};

function validateEmail(value: string) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(value),
    CHECK.isEmailValid(value),
  ]);
}

type FormData = {
  email: string;
};

function ServerLogin() {
  const [linkSent, setLinkSent] = useState(false);
  const { serverId } = useParams<RouteServerParams>();
  const { getServer } = useServers();
  const emailReceiver = useRef('');
  const server = getServer(serverId);

  const {
    handleSubmit,
    setValue,
    unregister,
    register,
    errors,
    setError,
  } = useForm<FormData>({ defaultValues: { email: '' } });

  useEffect(() => {
    register('email', { validate: validateEmail });
    return () => unregister('email');
  }, [register, unregister, setValue]);

  useEffect(() => {
    const onServerLoginReply = (
      _: unknown,
      { success }: ServerLoginResponse
    ) => {
      if (success) {
        setLinkSent(true);
      } else {
        setError('email', { message: 'Cannot not login to server' });
      }
    };

    ipcRenderer.on('serverLoginReply', onServerLoginReply);

    return () => {
      ipcRenderer.removeListener('serverLoginReply', onServerLoginReply);
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  function onSubmit({ email }: FormData) {
    if (server) {
      ipcRenderer.send('serverLogin', { serverId: server.id, email });
      emailReceiver.current = email;
    } else {
      console.error(`Cannot get server`);
    }
  }

  const actions = [
    <Button
      key="cancel"
      label="CANCEL"
      className={styles.cancelButton}
      to={ROUTE.HOME}
    />,
    <Button
      key="login"
      label="SEND ME A LOGIN LINK"
      className={styles.loginButton}
      onClick={handleSubmit(onSubmit)}
      primary
    />,
  ];

  return (
    <DefaultPage
      title="Login to Remote Server"
      subtitle="Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. "
      actions={actions}
    >
      <div className={styles.container}>
        <div className={styles.information}>
          <div className={styles.url}>
            <p className={styles.title}>SERVER URL</p>
            <p className={styles.value} title={server?.url}>
              {server?.url}
            </p>
          </div>
          <div className={styles.name}>
            <p className={styles.title}>SERVER NAME</p>
            <div className={styles.content}>
              <div className={styles.circle} />
              <p className={styles.value}>{server?.name}</p>
            </div>
          </div>
        </div>
        <div className={styles.form}>
          <div className={styles.formEmail}>
            <TextInput
              label="email"
              onChange={(value: string) => {
                setError('email', {});
                setValue('email', value);
              }}
              Icon={EmailIcon}
              onEnterKeyPress={handleSubmit(onSubmit)}
              error={errors.email?.message}
              showClearButton
              autoFocus
            />
          </div>
          <div className={cx(styles.confirmation, { [styles.show]: linkSent })}>
            <div className={styles.icon}>
              <EmailIcon className="icon-regular" />
            </div>
            <p>{`link sent to "${emailReceiver.current}". Check your inbox.`}</p>
          </div>
        </div>
      </div>
    </DefaultPage>
  );
}

export default ServerLogin;
