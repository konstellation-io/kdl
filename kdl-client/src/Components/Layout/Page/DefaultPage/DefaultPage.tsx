import React, { FC } from 'react';

import ActionsBar from 'Components/Layout/ActionsBar/ActionsBar';
import Sidebar from './Sidebar';
import styles from './DefaultPage.module.scss';

type Props = {
  title: string;
  subtitle?: string;
  children: JSX.Element;
  actions?: JSX.Element | JSX.Element[];
};
const DefaultPage: FC<Props> = ({ title, subtitle, actions, children }) => (
  <div className={styles.container}>
    <Sidebar title={title} subtitle={subtitle} />
    <div className={styles.content}>
      <div className={styles.children}>{children}</div>
      {actions && <ActionsBar className={styles.actions}>{actions}</ActionsBar>}
    </div>
  </div>
);

export default DefaultPage;
