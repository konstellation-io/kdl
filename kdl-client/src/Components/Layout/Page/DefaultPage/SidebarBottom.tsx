import React, { FC } from 'react';

import { bottomRef } from './Sidebar';
import { createPortal } from 'react-dom';
import styles from './DefaultPage.module.scss';

type Props = {
  children: JSX.Element;
};
const SidebarBottom: FC<Props> = ({ children }) => (
  <div className={styles.container}>
    {bottomRef?.current && createPortal(children, bottomRef.current)}
  </div>
);

export default SidebarBottom;
