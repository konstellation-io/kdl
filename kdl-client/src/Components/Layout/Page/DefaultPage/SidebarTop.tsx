import React, { FC } from 'react';

import { createPortal } from 'react-dom';
import styles from './DefaultPage.module.scss';
import { topRef } from './Sidebar';

type Props = {
  children: JSX.Element;
};
const SidebarTop: FC<Props> = ({ children }) => (
  <div className={styles.container}>
    {topRef?.current && createPortal(children, topRef.current)}
  </div>
);

export default SidebarTop;
