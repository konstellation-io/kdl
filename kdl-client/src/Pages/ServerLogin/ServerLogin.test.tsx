import { Button } from 'kwc';
import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import React from 'react';
import ServerLogin from './ServerLogin';
import { ipcRenderer } from 'electron';
import { shallow } from 'enzyme';

jest.mock('Hooks/useServers', () =>
  jest.fn(() => ({
    getServer: jest.fn(() => ({ serverUrl: 'some.url' })),
  }))
);

jest.mock('electron', () => ({
  ipcRenderer: {
    send: jest.fn(),
  },
}));

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(() => ({ serverId: 'serverId' })),
}));

// TODO: test useForm

const component = shallow(<ServerLogin />);

describe('ServerLogin', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    expect(component.contains('SERVER URL')).toBeTruthy();
  });

  test('action button', () => {
    component.find(DefaultPage).dive().find(Button).last().simulate('click');

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
  });
});
