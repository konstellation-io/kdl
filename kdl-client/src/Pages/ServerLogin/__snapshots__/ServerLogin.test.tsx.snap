// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerLogin Component match snapshot 1`] = `
<DefaultPage
  actions={
    Array [
      <P
        className="cancelButton"
        label="CANCEL"
        to="/"
      />,
      <P
        className="loginButton"
        label="SEND ME A LOGIN LINK"
        onClick={[Function]}
        primary={true}
      />,
    ]
  }
  subtitle="You will sign in to the a Remote Server. Make sure you have access privileges or you will not be able to complete the process. After introducing you email address, and email will allow you to sign in."
  title="Sign in to Remote Server"
>
  <div
    className="container"
  >
    <div
      className="information"
    >
      <div
        className="url"
      >
        <p
          className="title"
        >
          SERVER URL
        </p>
        <p
          className="value"
        />
      </div>
      <div
        className="name"
      >
        <p
          className="title"
        >
          SERVER NAME
        </p>
        <div
          className="content"
        >
          <div
            className="circle"
          />
          <p
            className="value"
          />
        </div>
      </div>
    </div>
    <div
      className="form"
    >
      <div
        className="formEmail"
      >
        <tt
          Icon={
            Object {
              "$$typeof": Symbol(react.memo),
              "compare": null,
              "displayName": "MailIcon",
              "muiName": "SvgIcon",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
            }
          }
          autoFocus={true}
          label="email"
          onChange={[Function]}
          onEnterKeyPress={[Function]}
          showClearButton={true}
        />
      </div>
      <div
        className="confirmation"
      >
        <div
          className="icon"
        >
          <MailIcon
            className="icon-regular"
          />
        </div>
        <p>
          link sent to "". Check your inbox.
        </p>
      </div>
    </div>
  </div>
</DefaultPage>
`;
